### PR TITLE
Clean up BeamRelevant with TypeMultiplier

### PR DESCRIPTION
**Pull request by n01r**
The `BeamRelevant` reduced diagnostics had a lot of code where `amrex::ReduceOps` were constructed from lists of types. This PR cleans it up with the `TypeMultiplier` utility that makes it more legible.

Credit goes to @AlexanderSinn's work in https://github.com/Hi-PACE/hipace/pull/1052 and https://github.com/AMReX-Codes/amrex/pull/3718.

### CODE REVIEW

File: Source/Diagnostics/ReducedDiags/BeamRelevant.cpp [hunk 1]
```diff
@@ -198,14 +199,14 @@ void BeamRelevant::ComputeDiags (int step)
 
         using PType = typename WarpXParticleContainer::SuperParticleType;
 
-        amrex::ReduceOps<ReduceOpSum,ReduceOpSum,ReduceOpSum,ReduceOpSum,ReduceOpSum,
-        ReduceOpSum,ReduceOpSum,ReduceOpSum> reduce_ops;
-        auto r = amrex::ParticleReduce<amrex::ReduceData<ParticleReal,ParticleReal,
-        ParticleReal,ParticleReal,ParticleReal,ParticleReal,ParticleReal,ParticleReal>>(
+        // number of reduction operations in first concurrent batch
+        const int num_red_ops_1 = 8;
```
Comment (AlexanderSinn) [inline Source/Diagnostics/ReducedDiags/BeamRelevant.cpp]:
```suggestion
        constexpr std::size_t num_red_ops_1 = 8;
```
Looks great! I would write constexpr explicitly here (and in the other places including in ImpactX). Currently, constexpr is implied by the const; for the code to function constexpr is necessary.

Comment (n01r) [reply Source/Diagnostics/ReducedDiags/BeamRelevant.cpp]: Thanks, @AlexanderSinn!

**Review by ax3l — APPROVED** [review]
Very nice :)

And maybe as a follow-up, we can make this a one-time pass reduction :)

File: Source/Diagnostics/ReducedDiags/BeamRelevant.cpp [hunk 2]
```diff
@@ -242,18 +243,17 @@ void BeamRelevant::ComputeDiags (int step)
...
+            reduce_ops_1);
+
+        std::vector<ParticleReal> values_per_rank_1st(num_red_ops_1);
+
+        /* contains in this order:
+         * w, x_mean, y_mean, z_mean
+         * ux_mean, uy_mean, uz_mean, gm_mean
+         */
+        amrex::constexpr_for<0, num_red_ops_1> ([&](auto i) {
+            values_per_rank_1st[i] = amrex::get<i>(r1);
```
Comment (ax3l) [inline Source/Diagnostics/ReducedDiags/BeamRelevant.cpp:255]: For some reason, @AlexanderSinn uses this syntax in HiPACE++:
```suggestion
            values_per_rank_1st[i.value] = amrex::get<i.value>(r1);
```

File: Source/Diagnostics/ReducedDiags/BeamRelevant.cpp [hunk 3]
```diff
@@ -336,19 +335,18 @@ void BeamRelevant::ComputeDiags (int step)
...
+
+        /* contains in this order:
+         * x_ms, y_ms, z_ms
+         * ux_ms, uy_ms, uz_ms,
+         * gm_ms
+         * xux, yuy, zuz,
+         * charge
+         */
+        amrex::constexpr_for<0, num_red_ops_2> ([&](auto i) {
+            values_per_rank_2nd[i] = amrex::get<i>(r2);
```
Comment (ax3l) [inline Source/Diagnostics/ReducedDiags/BeamRelevant.cpp:348]: see above

**Pull request merged**